code/datums/components/trapdoor.dm

```diff
@@ -0,0 +1,230 @@
+
+///makes this file more legible
+#define IS_OPEN(parent) isgroundlessturf(parent)
+///distance a trapdoor will accept a link request.
+#define TRAPDOOR_LINKING_SEARCH_RANGE 7
+
+/**
+ * # trapdoor component!
+ *
+ * component attached to floors to turn them into trapdoors, a constructable trap that when signalled drops people to the level below.
+ * assembly code at the bottom of this file
+ */
+/datum/component/trapdoor
+	/**
+	* list of lists that are arguments for readding decals when the floor comes back. pain.
+	*
+	* format: list(list(element's description, element's cleanable, element's directional, element's pic))
+	*/
+	var/list/stored_decals = list()
+	///assembly tied to this trapdoor
+	var/obj/item/assembly/trapdoor/assembly
+	///path of the turf this should change into when the assembly is pulsed. needed for openspace trapdoors knowing what to turn back into
+	var/trapdoor_turf_path
+
+/datum/component/trapdoor/Initialize(starts_open = FALSE, trapdoor_turf_path, assembly, given_decals)
```

Comment:
as far as i can tell this isn't called on the new trapdoor component. trapdoor_turf_path stays null, assembly is null as well.